### PR TITLE
Optical center relative to the image center

### PR DESCRIPTION
**Pull request by servantf**
replace principal point with offset from image center in camera internal description

### CODE REVIEW

**Comment by fabiencastan** [comment]
Need to add a compatibility code in IO files.

**Review by fabiencastan — CHANGES_REQUESTED** [review]
Need to add a compatibility code in IO files.

File: src/aliceVision/camera/IntrinsicsScaleOffset.hpp [hunk 1]
```diff
@@ -52,10 +54,19 @@ class IntrinsicsScaleOffset: public IntrinsicBase
     return _offset; 
   }
 
+  inline const Vec2 getPrincipalPoint() const
```
Comment (fabiencastan) [inline src/aliceVision/camera/IntrinsicsScaleOffset.hpp:57]:
```suggestion
  /**
   * @brief Principal point in image coordinate ((0,0) is image top-left).
   */
  inline const Vec2 getPrincipalPoint() const 
```

File: src/aliceVision/version.hpp [hunk 2]
```diff
@@ -19,3 +21,24 @@
...
+        }
+
+        if (version[i] > base[i])
+        {
+            return false;
+        }
+    }    
+
+    return false;
+}
```
Comment (fabiencastan) [inline src/aliceVision/version.hpp:42]: class Version
{
public:
    Version(const Vec3&);
    Version(int major, int minor, int micro);
    bool operator<(const Vec3& other) const;
private:
    Vec3 _v;
};

File: src/aliceVision/sfm/BundleAdjustmentCeres.cpp [hunk 3]
```diff
@@ -656,14 +665,14 @@ void BundleAdjustmentCeres::addIntrinsicsToProblem(const sfmData::SfMData& sfmDa
...
+    if((refineOptions & REFINE_INTRINSICS_OPTICALOFFSET_ALWAYS) ||
+       ((refineOptions & REFINE_INTRINSICS_OPTICALOFFSET_IF_ENOUGH_DATA) && _minNbImagesToRefineOpticalCenter > 0 && usageCount >= _minNbImagesToRefineOpticalCenter))
     {
       // refine optical center within 10% of the image size.
       assert(intrinsicBlock.size() >= 3);
 
-      const double opticalCenterMinPercent = 0.45;
-      const double opticalCenterMaxPercent = 0.55;
+      const double opticalCenterMinPercent = 0.45 - 0.5;
+      const double opticalCenterMaxPercent = 0.55 - 0.5;
```
Comment (fabiencastan) [inline src/aliceVision/sfm/BundleAdjustmentCeres.cpp:675]:
```suggestion
      const double opticalCenterMinPercent = -0.05;
      const double opticalCenterMaxPercent =  0.05;
```

File: src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp [hunk 4]
```diff
@@ -519,14 +519,14 @@ void BundleAdjustmentPanoramaCeres::addIntrinsicsToProblem(const sfmData::SfMDat
...
+    bool optional_center = ((refineOptions & REFINE_INTRINSICS_OPTICALOFFSET_IF_ENOUGH_DATA) && (usageCount > minImagesForOpticalCenter));
+    if((refineOptions & REFINE_INTRINSICS_OPTICALOFFSET_ALWAYS) || optional_center)
     {
       // refine optical center within 10% of the image size.
       assert(intrinsicBlock.size() >= 4);
 
-      const double opticalCenterMinPercent = 0.45;
-      const double opticalCenterMaxPercent = 0.55;
+      const double opticalCenterMinPercent = 0.45 - 0.5;
+      const double opticalCenterMaxPercent = 0.55 - 0.5;
```
Comment (fabiencastan) [inline src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp:529]:
```suggestion
      const double opticalCenterMinPercent = - 0.05;
      const double opticalCenterMaxPercent =  0.05;
```

File: src/aliceVision/sfmDataIO/sceneSample.cpp [hunk 5]
```diff
@@ -29,8 +29,8 @@ void generateSampleScene(sfmData::SfMData & output)
     const int h = 2048;
     const double focalLengthPixX = 1000.0;
     const double focalLengthPixY = 2000.0;
-    const double ppx = 2020.0;
-    const double ppy = 1040.0;
+    const double ppx = 2020.0 - (double(w) / 2.0);
+    const double ppy = 1040.0 - (double(h) / 2.0);
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/sceneSample.cpp:33]:
```suggestion
    const double ppx = -26;
    const double ppy = 16;
```

File: src/aliceVision/sfmDataIO/viewIO.cpp [hunk 6]
```diff
@@ -140,8 +140,8 @@ std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(
 
   camera::EINTRINSIC intrinsicType = defaultIntrinsicType;
 
-  double ppx = view.getWidth() / 2.0;
-  double ppy = view.getHeight() / 2.0;
+  double ppx = defaultPPx;
+  double ppy = defaultPPy;
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/viewIO.cpp:144]: The default value was -1 (in the hpp) to say "no value", now -1 is a valid value.

**Review by fabiencastan — CHANGES_REQUESTED** [review]

File: src/aliceVision/camera/IntrinsicsScaleOffset.hpp [hunk 7]
```diff
@@ -152,6 +167,22 @@ class IntrinsicsScaleOffset: public IntrinsicBase
     return true;
   }
 
+  bool importFromParams(const std::vector<double>& params, const Vec3 & inputVersion) override
```
Comment (fabiencastan) [inline src/aliceVision/camera/IntrinsicsScaleOffset.hpp:170]:
```suggestion
  /**
   * @brief Import a vector of params loaded from a file. It is similar to updateFromParams but it deals with file compatibility.
   */
  bool importFromParams(const std::vector<double>& params, const Vec3 & inputVersion) override
```

File: src/aliceVision/version.hpp [hunk 8]
```diff
@@ -19,3 +21,40 @@
...
+            {
+                return false;
+            }
+        }    
+
+        return false;
+    }
+
+private:
+    Vec3 _v;
```
Comment (fabiencastan) [inline src/aliceVision/version.hpp:57]:
```suggestion
    Vec3i _v;
```

Comment (servantf) [reply src/aliceVision/version.hpp:57]: It's obvious that integers are better suited for this task.
However, i choose doubles as Json I/O has always loaded/saved doubles.

File: src/aliceVision/sfmDataIO/CHANGES.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# SfmDataIO Changelog
+
+List of changes to the file formats ABC (Alembic) and SFM (JSON).
+
+
+## Develop Version
+
+### File Version 1.2.1
+- The principal point (or optical center) is now relative to the center of image (and no more to the top-left corner). It is defined in pixel coordinates in all cases.
```
Comment (simogasp) [inline src/aliceVision/sfmDataIO/CHANGES.md:9]:
```suggestion
- The principal point is now relative to the center of image (and no more to the top-left corner). It is defined in pixel coordinates in all cases.
```
The principal point is the projection of the optical center

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**